dipy/reconst/csdeconv.py

```diff
@@ -396,48 +483,52 @@ def csdeconv(dwsignal, sh_order, X, B_reg, lambda_=1., tau=0.1):
            constrained super-resolved spherical deconvolution.
 
     """
-    # generate initial fODF estimate, truncated at SH order 4
-    fodf_sh = np.linalg.lstsq(X, dwsignal)[0]
-    fodf_sh[15:] = 0
-
-    fodf = np.dot(B_reg, fodf_sh)
+    mu = 1e-5
+    if P is None:
+        P = np.dot(X.T, X)
+    z = np.dot(X.T, dwsignal)
+
+    try:
+        fodf_sh = _solve_cholesky(P, z)
+    except la.LinAlgError:
+        P = P + mu * np.eye(P.shape[0])
```

Comment:
Well ... P is an argument to the function so doing this in place will change P for the caller.

We could handle this outside of the function all together, but than I think we would need to make P a required argument instead of an optional one.